mi/match.c

```diff
@@ -163,564 +180,580 @@ mkdtree(Srcloc loc, Node *lbl)
+	 * c points to the slot at the path pi.
+	 */
+	c = NULL;
+	slot = NULL;
+	nslot = 0;
+	for (i = 0; i < fs->nslot; i++) {
+		if (patheq(pi, fs->slot[i]->path)) {
+			c = fs->slot[i];
+			continue;
 		}
+		lappend(&slot, &nslot, fs->slot[i]);
 	}
 
-	if (!start->load) {
-		start->load = utag(val);
-		start->nconstructors = nconstructors(tybase(exprtype(pat)));
-	}
-
-	tagid = mkintlit(pat->loc, uc->id);
-	tagid->expr.type = mktype(pat->loc, Tyint32);
-	lappend(&start->pat, &start->npat, tagid);
-	if (uc->etype) {
-		next = mkdtree(pat->loc, genlbl(pat->loc));
-		lappend(&start->next, &start->nnext, next);
-		addpat(pat->expr.args[1], uvalue(val, uc->etype), next, accept, cap, ncap, end, nend);
-	} else {
-		lappend(&start->next, &start->nnext, accept);
-		lappend(end, nend, accept);
+	out = zalloc(sizeof(Frontier));
+	out->i = fs->i;
+	out->lbl = fs->lbl;
+	out->slot = slot;
+	out->nslot = nslot;
+	out->cap = fs->cap;
+	out->ncap = fs->ncap;
+	out->final = fs->final;
+
+	/*
+	 * if the sub-term at pi is not in the frontier,
+	 * then we do not reduce the frontier.
+	 */
+	if (c == NULL)
+		return out;
+
+	switch (exprop(c->pat)) {
+	case Ovar:
+	case Ogap:
+		/*
+		 * if the pattern at the sub-term pi of this frontier is not a constructor,
+		 * then we do not reduce the frontier.
+		 */
+		return out;
+	default:
+		break;
 	}
-	return 1;
-}
-
-static int
-addstr(Node *pat, Node *val, Dtree *start, Dtree *accept, Node ***cap, size_t *ncap, Dtree ***end, size_t *nend)
-{
-	Dtree **tail, **last, *next;
-	size_t i, j, n, ntail, nlast;
-	Node *p, *v, *lit;
-	Type *ty;
-	char *s;
-	int ret;
 
-	lit = pat->expr.args[0];
-	n = lit->lit.strval.len;
-	s = lit->lit.strval.buf;
-
-	ty = mktype(pat->loc, Tyuint64);
-	p = mkintlit(lit->loc, n);
-	v = structmemb(val, mkname(pat->loc, "len"), ty);
-	p->expr.type = ty;
-
-	if (n == 0)
-		next = accept;
-	else
-		next = mkdtree(pat->loc, genlbl(pat->loc));
+	/*
+	 * if constructor at the path pi is not the constructor we want to project,
+	 * then return null.
+	 */
+	if (!pateq(pat, c->pat))
+		return NULL;
 
-	last = NULL;
-	nlast = 0;
-	ret = 0;
-	if (addpat(p, v, start, next, cap, ncap, &last, &nlast))
-		ret = 1;
-
-	ty = mktype(pat->loc, Tybyte);
-	for (i = 0; i < n; i++) {
-		p = mkintlit(lit->loc, s[i]);
-		p->expr.type = ty;
-		v = arrayelt(val, i);
-
-		tail = NULL;
-		ntail = 0;
-		next = nextnode(pat->loc, i, n, accept);
-		for (j = 0; j < nlast; j++)
-			if (addpat(p, v, last[j], next, NULL, NULL, &tail, &ntail))
-				ret = 1;
-		lfree(&last, &nlast);
-		last = tail;
-		nlast = ntail;
-	}
-	lcat(end, nend, last, nlast);
-	lfree(&last, &nlast);
-	return ret;
+	return out;
 }
 
-static int
-addlit(Node *pat, Node *val, Dtree *start, Dtree *accept, Node ***cap, size_t *ncap, Dtree ***end, size_t *nend)
+/* compile implements the algorithm outlined in the paper "When Do Match-Compilation Heuristics Matter?" by Kevin Scott and Norman Ramsey
```

Comment:
Some of these lines are a bit long. I don't go for hard character limits (it's better to go over by a few chars than to wrap things oddly), but if you could wrap this comment, that would be appreciated.

Reply:
Done